File: day06/aws-vpc-python/cidr_subnet_utils.py
from ipaddress import ip_network, IPv4Address


def split_subnets(base_cidr, new_prefix_length):
    subnets = []
    base_subnet = ip_network(base_cidr)
    first_address = int(base_subnet.network_address)
    last_address = int(base_subnet.broadcast_address)
    mask = ip_network(f"0.0.0.0/{new_prefix_length}")
    increment = int(mask.netmask) - int(IPv4Address('255.255.255.255'))

    i = first_address
    while i <= last_address:
        subnet = f"{IPv4Address(i)}/{new_prefix_length}"
        subnets.append(subnet)
        i += abs(increment) + 1

    return subnets

File: day13/aws-vpc-ts/cidr_subnet_utils.ts
import * as ip from 'ip';

export function splitSubnets(baseCidr: string, newPrefixLength: number): string[] {
  let subnets: string[] = [];
  let baseSubnet = ip.cidrSubnet(baseCidr);
  let firstAddress = ip.toLong(baseSubnet.firstAddress);
  let lastAddress = ip.toLong(baseSubnet.lastAddress);
  let mask = ip.fromPrefixLen(newPrefixLength);
  let increment = ip.toLong(mask) - ip.toLong('255.255.255.255');

  for (let i = firstAddress; i <= lastAddress; i += Math.abs(increment) + 1) {
    let subnet = ip.fromLong(i - 1) + '/' + newPrefixLength;
    subnets.push(subnet);
  }

  return subnets;
}
